external-crates/move/crates/move-abstract-interpreter/src/absint.rs

```diff
@@ -1,228 +1,151 @@
-// Copyright (c) The Diem Core Contributors
 // Copyright (c) The Move Contributors
 // SPDX-License-Identifier: Apache-2.0
 
-use crate::control_flow_graph::{BlockId, ControlFlowGraph, VMControlFlowGraph};
-use move_binary_format::{
-    errors::PartialVMResult,
-    file_format::{
-        AbilitySet, Bytecode, CodeOffset, CodeUnit, FunctionDefinitionIndex, FunctionHandle,
-        Signature,
-    },
-    CompiledModule,
-};
-use move_bytecode_verifier_meter::{Meter, Scope};
+use crate::control_flow_graph::ControlFlowGraph;
 use std::collections::BTreeMap;
 
-/// A `FunctionContext` holds all the information needed by the verifier for `FunctionDefinition`.`
-/// A control flow graph is built for a function when the `FunctionContext` is created.
-pub struct FunctionContext<'a> {
-    index: Option<FunctionDefinitionIndex>,
-    code: &'a CodeUnit,
-    parameters: &'a Signature,
-    return_: &'a Signature,
-    locals: &'a Signature,
-    type_parameters: &'a [AbilitySet],
-    cfg: VMControlFlowGraph,
-}
-
-/// Trait for finite-height abstract domains. Infinite height domains would require a more complex
-/// trait with widening and a partial order.
-pub trait AbstractDomain: Clone + Sized {
-    fn join(
-        &mut self,
-        other: &Self,
-        meter: &mut (impl Meter + ?Sized),
-    ) -> PartialVMResult<JoinResult>;
-}
-
 #[derive(Debug)]
 pub enum JoinResult {
     Changed,
     Unchanged,
 }
 
-#[allow(dead_code)]
-#[derive(Clone)]
-pub struct BlockInvariant<State> {
-    /// Precondition of the block
-    pre: State,
-}
-
-/// A map from block id's to the pre/post of each block after a fixed point is reached.
-#[allow(dead_code)]
-pub type InvariantMap<State> = BTreeMap<BlockId, BlockInvariant<State>>;
+pub trait AbstractInterpreter {
+    type Error;
+    type BlockId: Copy + Ord;
 
-/// Costs for metered verification
-const ANALYZE_FUNCTION_BASE_COST: u128 = 10;
-const EXECUTE_BLOCK_BASE_COST: u128 = 10;
-const PER_BACKEDGE_COST: u128 = 10;
-const PER_SUCCESSOR_COST: u128 = 10;
+    type State: Clone;
+    type InstructionIndex: Copy + Ord;
+    type Instruction;
 
-/// Take a pre-state + instruction and mutate it to produce a post-state
-/// Auxiliary data can be stored in self.
-pub trait TransferFunctions {
-    type State: AbstractDomain;
-    type Error;
+    fn start(&mut self) -> Result<(), Self::Error>;
+    fn join(
```

Comment:
I suspect that for fancier (non bytecode verifier) analyses, we are going to need `fn widen(&mut self, pre: &mut Self::State, post: &Self::State)` (or similar) as well.

Reply:
It is a good callout, and I think it is worth adding once we need it. For now though I hesitate mostly around adding it just in that we don't really know what shape it will take. 
In other words, if we had a pass that we wanted to implement with widening, it would help inform the API a lot